tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause.bal

```diff
@@ -357,6 +360,138 @@ function getError() returns error {
     return err;
 }
 
+public class Listener {
+
+    public isolated function 'start() returns error? {
+        return externStart(self);
+    }
+    public isolated function gracefulStop() returns error? {
+    }
+    public isolated function immediateStop() returns error? {
+    }
+    public isolated function detach(service object {} s) returns error? {
+    }
+    public isolated function attach(service object {} s, string[]|string? name = ()) returns error? {
+        return self.register(s, name);
+    }
+    isolated function register(service object {} s, string[]|string? name) returns error? {
+        return externAttach(s, name);
+    }
+
+    public function init() returns error? {
+        check externLInit(self);
+    }
+}
+
+type S service object {
+};
+
+service S / on new Listener() {
+    resource isolated function get foo() returns service object {} {
+        return service object {
+            resource function get foo() returns string {
+                do {
+                    _ = check getCheckError();
+                    return "should not reach here";
+                } on fail error e {
+                    return "Error thrown from on-fail: " + e.message();
+                }
+            }
+        };
+    }
+
+    resource isolated function get bar() returns service object {} {
+        return service object {
+            resource function get bar() returns string {
+                do {
+                    do {
+                        _ = check getCheckError();
+                    } on fail error e1 {
+                        fail e1;
+                    }
+                } on fail error e2 {
+                    return "Error thrown from on-fail: " + e2.message();
+                }
+                return "should not reach here";
+            }
+        };
+    }
+}
+
+function getService() returns object {} = @java:Method {
+    'class: "org/ballerinalang/nativeimpl/jvm/servicetests/ServiceValue",
```

Comment:
@rdhananjaya is it okay to use the same `ServiceValue` in this new test case? I slightly remember you said there could be a conflict. So need to use a new one?

Reply:
The same `ServiceValue` is being atm in different tests atm though

1. https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/test-src/services/service_decl.bal
2. https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/test-src/bala/test_bala/object/test_service_objects.bal
3. https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/test-src/klass/simple_service_class.bal)

Reply:
ServiceValue class is a static one so the test cases need to make sure that tests that use the ServiceValue do not run concurrently and make sure that ServiceValue is reset before and after the test.

Reply:
Each pkg in jbal-unit-tests runs consecutively (https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/testng.xml#L24). These tests are under a separate pkg(https://github.com/pcnfernando/ballerina-lang/blob/2201.1.x/tests/jballerina-unit-test/src/test/resources/testng.xml#L105). `ServiceValue` is reset at the beginning and end. Hence, should be fine then. Thanks